pyocd/probe/stlink_probe.py

```diff
@@ -329,8 +329,8 @@ def read_memory_block8(self, addr: int, size: int, **attrs: Any) -> Sequence[int
         res = []
 
         # read leading unaligned bytes
-        unaligned_count = addr & 3
-        if (size > 0) and (unaligned_count > 0):
+        unaligned_count = 3 & (4 - (addr & 3))
+        if (size > unaligned_count > 0):
```

Comment:
This still needs to read the unaligned bytes even if `unaligned_count` is less than or equal to `size`. So this should be:

```py
unaligned_count = min(size, 3 & (4 - addr))
if unaligned_count > 0:
```

(The `addr & 3` mask is unnecessary given how integers and bit masking works in Python.)

Reply:
It looks fine.

Reply:
Thanks for the comments. 👍

I do agree the unaligned read still has to happen, but not necessarily here (in case of a small `size` value).

Please note that  `size <= unaligned_count` also means `size < 4`, so the second if statement is also skipped making the very last read (line 347) handle the whole (short) request. Please see the commit message:

> Skip the first unaligned transfer if the requested size is smaller than
the result of the address alignment check. Handle the whole request in
the second, unaligned read/write.

Reply:
If you think the request should be always handled within the first conditional, or that https://github.com/pyocd/pyOCD/pull/1476/commits/a353c96745559105dddf711bd489057da55a1c88#r1032679727 code is more readable, please let me know, so I can amend the patch this way.

Reply:
Ooh, totally missed that, really sorry!! Thanks for explaining it. (And I didn't see the commit message since GH collapses all but the commit subject by default.)

It would probably be best to have a comment describing how this works before this section of code.
Whether you keep the (strictly unnecessary) `addr & 3` bit is your choice.

Reply:
Done; I've removed the redundant `& 3` operation and added a comment in the code.